atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/ring/RingGraph.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.workload.workflow.ring;
+
+import com.google.common.collect.Iterators;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import one.util.streamex.EntryStream;
+import one.util.streamex.IntStreamEx;
+
+public final class RingGraph {
```

Comment:
Actually, looking at this a little more, I see we use `generateNewRing`. Normally, this should only be called on a *verified* ring. To make illegal states unrepresentable, we could have this,
- `Graph` something which represents all states we have
- `Ring`
- `graph.toRingOrError()` which returns some union type
- `Ring` should be the one to have `generateNewRing` and not `Graph`

This is because I really like shuffling on each new ring - good for the RNG quality!

Again, I don't feel strongly about this

Reply:
Happy to re-name `validate` to `checkForErrors` if that's easier!

 Yeah agree that the naming and arguments are frustrating here... my only hesitation with making a general purpose `graph` class, is that the `toRing` or something equivalent wouldn't really make sense, as it's too specific for our use-case. I did think of maybe making `RingGraph` and `PartialRingGraph` (or `UnverifiedRingGraph`), which would then perform the checks/etc, but thought it might be too overkill. 

As for `Normally, this should only be called on a verified ring`, this is not necessarily the case; we call it when we have an empty ring (on first write), or after we hit an error (so we don't stop the experiment, and continue writing). That said, I think this is just simply too much cost involved with trying to handle this partial states and errors, I'm going to opt for we just stop the workflow if we detect an error. This then will greatly simplify the interface for the class.